sample/volumeRenderingTexture3D/main.ts

```
@@ -7,21 +7,51 @@ const canvas = document.querySelector('canvas') as HTMLCanvasElement;
 
 const gui = new GUI();
 
+const brainImages = {
+  r8unorm: {
+    bytesPerBlock: 1,
+    blockLength: 1,
+    dataPath:
+      '../../assets/img/volume/t1_icbm_normal_1mm_pn0_rf0_180x216x180_uint8_1x1.bin-gz',
+  },
+  'bc4-r-unorm': {
+    bytesPerBlock: 8,
+    blockLength: 4,
+    dataPath:
+      '../../assets/img/volume/t1_icbm_normal_1mm_pn0_rf0_180x216x180_bc4_4x4.bin-gz',
+  },
+};
+
 // GUI parameters
-const params: { rotateCamera: boolean; near: number; far: number } = {
+const params: {
+  rotateCamera: boolean;
+  near: number;
+  far: number;
+  textureFormat: GPUTextureFormat;
+} = {
   rotateCamera: true,
   near: 2.0,
   far: 7.0,
+  textureFormat: 'r8unorm',
 };
 
 gui.add(params, 'rotateCamera', true);
 gui.add(params, 'near', 2.0, 7.0);
 gui.add(params, 'far', 2.0, 7.0);
+gui.add(params, 'textureFormat', Object.keys(brainImages)).onChange(() => {
+  createVolumeTexture(params.textureFormat);
```

On Chrome Stable (and presumably on Canary on hardware that doesn't support BC) I get this when I switch:

> Uncaught (in promise) TypeError: Failed to execute 'createTexture' on 'GPUDevice': Use of the 'bc4-r-unorm' texture format requires the 'texture-compression-bc' feature to be enabled on [Device].

This sample should do something to gracefully fall back (at least, don't throw an error and then keep showing the r8unorm image).

Also, maybe the Safari issue is that it's failing silently instead of throwing an exception?

---

Instead of `Object.keys(brainImages)`, we could use `device.features` to selectively add only the texture formats that we know will work. However, I've kept all of them initially so that users can see that the lack of device features is why it's not working: WebGPU supports BC and ASTC but not this machine.

---

That's fine, I do think we should list all of the formats supported by the sample even on systems that don't support them. The problem is if I select an unsupported format (like BC on Chrome Stable right now) then it silently fails to update the texture, then keeps rendering with the r8unorm version. There's no indication that it failed unless I open dev tools.

---

I'm surprised `device.onuncapturederror` is not triggered.
https://github.com/webgpu/webgpu-samples/blob/85e141cd46bad0e69982f2baab1bc4f4b6eccb22/sample/util.ts#L49-L51

I'll have a look tomorrow.

---

Oh that's because `createTexture` does not generate a validation error, but a typeerror! I'll fix it.

---

That's much better indeed! Thank you @kainino0x

---

FYI next to come is https://github.com/beaufortfrancois/webgpu-samples/compare/bc4...beaufortfrancois:webgpu-samples:astc12?expand=1

---

I've updated the code to stop animation when downloading images to make things clearer.

---

I am still seeing an unhandled promise rejection when I try this in Chrome Stable. I don't know what we should show, but not that, because it says "unhandled promise rejection, please report a bug!"

---

Pushed a commit to fix this